iothub/device/src/Authentication/IotHubConnectionStringBuilder.cs

```diff
@@ -57,74 +57,50 @@ internal IotHubConnectionStringBuilder()
+        /// <param name="gatewayHostName">The fully-qualified DNS host name of the gateway.</param>
         /// <returns>A new instance of the <see cref="IotHubConnectionStringBuilder"/> class with a populated connection string.</returns>
-        public static IotHubConnectionStringBuilder Create(string hostname, string gatewayHostname, IAuthenticationMethod authenticationMethod)
+        public IotHubConnectionStringBuilder(IAuthenticationMethod authenticationMethod, string hostName, string gatewayHostName = null)
         {
-            var iotHubConnectionStringBuilder = new IotHubConnectionStringBuilder
-            {
-                HostName = hostname,
-                GatewayHostName = gatewayHostname,
-                AuthenticationMethod = authenticationMethod,
-            };
-
-            iotHubConnectionStringBuilder.Validate();
+            AuthenticationMethod = authenticationMethod;
+            HostName = hostName;
+            GatewayHostName = gatewayHostName;
 
-            return iotHubConnectionStringBuilder;
+            Validate();
         }
 
         /// <summary>
         /// Creates a connection string based on the hostname of the IoT hub and the authentication method passed as a parameter.
         /// </summary>
         /// <param name="iotHubConnectionString">The connection string.</param>
         /// <returns>A new instance of the <see cref="IotHubConnectionStringBuilder"/> class with a populated connection string.</returns>
-        public static IotHubConnectionStringBuilder Create(string iotHubConnectionString)
+        public IotHubConnectionStringBuilder(string iotHubConnectionString)
+            : this(iotHubConnectionString, null)
         {
-            if (iotHubConnectionString.IsNullOrWhiteSpace())
-            {
-                throw new ArgumentNullException(nameof(iotHubConnectionString));
-            }
-
-            return CreateWithIAuthenticationOverride(iotHubConnectionString, null);
         }
 
-        internal static IotHubConnectionStringBuilder CreateWithIAuthenticationOverride(
-            string iotHubConnectionString,
-            IAuthenticationMethod authenticationMethod)
+        internal IotHubConnectionStringBuilder(string iotHubConnectionString, IAuthenticationMethod authenticationMethod)
         {
-            var csBuilder = new IotHubConnectionStringBuilder
-            {
-                HostName = "TEMP.HUB",
-            };
+            Argument.AssertNotNullOrWhiteSpace(iotHubConnectionString, nameof(iotHubConnectionString));
 
             if (authenticationMethod == null)
             {
-                csBuilder.Parse(iotHubConnectionString);
-                csBuilder.AuthenticationMethod = AuthenticationMethodFactory.GetAuthenticationMethod(csBuilder);
+                // We'll parse the connection string and use that to build an auth method
+                Parse(iotHubConnectionString);
```

Comment:
This class always read a bit weird to me because of the way the constructor, `Create` and `Parse` methods were used. I'd expect a `Parse` method to return the "parsed" object, but that was weirdly managed by this class because the `Parse` method set member variables. Now that you've refactored the "Create" logic, is there a way for us to modify the "Parse" method as well so that we either set member variables in the constructor or have the `Parse` method return the "parsed" object.
Do you think that is something this class can allow?

Reply:
Parse is just a private method, so I don't really compare it to Create. It isn't as if users of the class would see both Create and Parse as options.

The refactored Create logic was just getting rid of the factory method and using constructors instead.

If we made parse only get (not set) the values, it would need some (new) object to put them on, then the ctor would then assign the properties from that other object. Honestly, I don't see the point of that.

To me, the weirdest part of this whole thing is that we create an auth method and pass `this`. The new auth method object now changes properties on this class. That was unexpected and creates the 2-ways-to-create-this-object.

So, looking at it again, I decided to split the cs + authmethod into 2 methods and found that we only had references for cs only. So, I took out the cs + authmethod overload. That reduces that oddness.